tools/CorrelationTestbed/InstallAndCheckCorrelation/InstallAndCheckCorrelation/InstallAndCheckCorrelation.cpp

```diff
@@ -0,0 +1,552 @@
+    {
+        hr = E_INVALIDARG;
+        error = "A source name must be supplied, use -src";
+        goto output_result;
+    }
+
+    // Execute the install step
+    phase = "Install";
+    std::cout << "Connecting to PackageManager..." << std::endl;
+    try
+    { do {
+        action = "Create package manager";
+        auto packageManager = CreatePackageManager(useDevCLSIDs);
+
+        action = "Get source reference";
+        auto sourceRef = packageManager.GetPackageCatalogByName(ConvertToUTF16(sourceName));
+
+        action = "Connecting to catalog";
+        auto connectResult = sourceRef.Connect();
+
+        if (connectResult.Status() != ConnectResultStatus::Ok)
+        {
+            hr = E_FAIL;
+            error = "Error connecting to catalog";
+            break;
+        }
+
+        auto catalog = connectResult.PackageCatalog();
+
+        action = "Create find options";
+        auto findOptions = CreateFindPackagesOptions(useDevCLSIDs);
+
+        action = "Add package id filter";
+        auto filter = CreatePackageMatchFilter(useDevCLSIDs);
+        filter.Field(PackageMatchField::Id);
+        filter.Option(PackageFieldMatchOption::Equals);
+        filter.Value(ConvertToUTF16(packageIdentifier));
+        findOptions.Filters().Append(filter);
+
+        action = "Find package";
+        auto findResult = catalog.FindPackages(findOptions);
+
+        if (findResult.Status() != FindPackagesResultStatus::Ok)
+        {
+            hr = E_FAIL;
+            error = "Error finding packages";
+            break;
+        }
+
+        action = "Get match";
+        auto matches = findResult.Matches();
+
+        if (matches.Size() == 0)
+        {
+            hr = E_NOT_SET;
+            error = "Package not found";
+            break;
+        }
+
+        auto package = matches.GetAt(0).CatalogPackage();
+
+        action = "Inspect package";
+        auto installVersion = package.DefaultInstallVersion();
+        packageName = ConvertToUTF8(installVersion.DisplayName());
+        if (useDevCLSIDs)
+        {
+            // Publisher is not yet available on the release version; make this unconditional when it is
+            packagePublisher = ConvertToUTF8(installVersion.Publisher());
+        }
+
+        action = "Create install options";
+        auto installOptions = CreateInstallOptions(useDevCLSIDs);
+
+        installOptions.PackageInstallScope(PackageInstallScope::Any);
+        installOptions.PackageInstallMode(PackageInstallMode::Silent);
+
+        std::cout << "Beginning to install " << packageIdentifier << " (" << packageName << ") from " << sourceName << "..." << std::endl;
+        auto installResult = packageManager.InstallPackageAsync(package, installOptions).get();
+
+        if (installResult.Status() != InstallResultStatus::Ok)
+        {
+            hr = installResult.ExtendedErrorCode();
+            error = "Error installing package";
+            break;
+        }
+
+    } while(0); }

```

Comment:
I don't have any actual suggestion, but using the do-while + break like this just doesn't feel right to me..